vernac/classes.ml

```diff
@@ -152,6 +152,28 @@ let default_locality () =
 let instance_locality =
   Attributes.hint_locality ~default:default_locality
 
+type event =
+  | NewClass of typeclass
+  | NewInstance of {
+      class_name : GlobRef.t;
+      instance : GlobRef.t;
+      info : Typeclasses.hint_info;
+      locality : Hints.hint_locality;
+    }
+
+let observers = ref CString.Map.empty
+
+let add_observer ~name o = observers := CString.Map.add name o !observers
+
+let observe event =
+  CString.Map.iter (fun name f ->
+    try f event
+    with e when CErrors.noncritical e ->
+      let e = Exninfo.capture e in
+      CErrors.user_err Pp.(str"Typeclass observer " ++ str name ++ str ":" ++ fnl () ++
```

Comment:
This should probably be a distinct error ie
~~~ocaml
exception ObserveError of string * exception

let () = register_printer ...

let observe event = 
  ...
  with e when noncritical e -> let e, info = capture e in iraise ~info (ObserveError (name, e))
  ...
~~~

Reply:
done